torch/csrc/jit/passes/pass_manager.h

```diff
@@ -23,17 +23,77 @@ namespace jit {
 // A pass modifies a Graph in place.
 using GraphPass = std::function<void(std::shared_ptr<Graph>&)>;
 
-TORCH_API std::vector<GraphPass>& getCustomPostFusionPasses();
-TORCH_API std::vector<GraphPass>& getCustomPreFusionPasses();
+// Since Passes are std::functions, we associate a UUID to each pass, this way
+// if we want to deregister a pass, we have something to reference it by.
+using GraphPassNameType = unsigned int;
+// Start UUID at 1
+static GraphPassNameType graphPassID = 1;
 
-struct TORCH_API RegisterPostFusionPass {
-  RegisterPostFusionPass(GraphPass p);
-};
+// Graph pass entries have a name associated with them
+using GraphPassEntry = std::pair<GraphPass, GraphPassNameType>;
+
+// Return currently registered passes. Passes are stored in a static vector
+TORCH_API std::vector<std::pair<GraphPass, GraphPassNameType>>& getCustomPostPasses();
+TORCH_API std::vector<std::pair<GraphPass, GraphPassNameType>>& getCustomPrePasses();
+
+TORCH_API GraphPassNameType registerPostPass(GraphPass p);
+TORCH_API GraphPassNameType registerPrePass(GraphPass p);
+//Default to registerPostPass
+TORCH_API GraphPassNameType registerPass(GraphPass p);
+
+// Look up pass by name passed in, remove it from registered passes
+TORCH_API void clearPostPass(GraphPassNameType p);
+TORCH_API void clearPrePass(GraphPassNameType p);
+
+// Remove all passes
+TORCH_API void clearAllPostPasses();
+TORCH_API void clearAllPrePasses();
+
+/*
+ * PassManager is a wrapper on the register/clear PostPass functions above. It
+ * will register the pass provided in "registerPass" and will hold on to its
+ * associated name that way clearPass can be later called and will delete the
+ * pass used to register when called.
+ *
+ * PassManager is templated because we want static variables based on a
+ * particular GraphPass. When deriving from PassManager, you should send as the
+ * template parameter your derived class as you would for the curiously
+ * recurring template pattern. This template parameter isn't actually used and
+ * is simply done to prevent static members from being shared across derived
+ * types.
+ */
+template <typename DerivedType>
+struct TORCH_API PassManager {
+ private:
+  // We want this class to be abstract because it's
+  virtual void abstract() = 0;
 
-using RegisterPass = RegisterPostFusionPass;
+ protected:
+  /*
+   * isRegistered() will return if a pass has been registered
+   * isRegistered(true) will change the value of the internal static bool
+   *
+   * There's an internal static bool to this function to keep track of the
+   * state, this is so when functions are derived from this class, they don't
+   * have to worry about initializing the static members.
+   */
+  static bool isRegistered(bool flip_bit = false);
```

Comment:
Can we make `isRegistered` a static member of this class and add getters/setters for it instead of one function that does both?